### PR TITLE
Check for a nil lastTime before use

### PR DESCRIPTION
**Pull request by Nalum**
Fix: #246 

Avoiding a nil-deref by checking the `lastTime` variable before using it.

### CODE REVIEW

**Comment by Nalum** [comment]
Not too sure about how to test this, happy to get some pointers.

File: controllers/imageupdateautomation_controller.go [hunk 1]
```diff
@@ -319,7 +319,7 @@ func (r *ImageUpdateAutomationReconciler) Reconcile(ctx context.Context, req ctr
 		if err == errNoChanges {
 			debuglog.Info("no changes made in working directory; no commit")
 			statusMessage = "no updates made"
-			if lastCommit, lastTime := auto.Status.LastPushCommit, auto.Status.LastPushTime; lastCommit != "" {
+			if lastCommit, lastTime := auto.Status.LastPushCommit, auto.Status.LastPushTime; lastCommit != "" && lastTime != nil {
 				statusMessage = fmt.Sprintf("%s; last commit %s at %s", statusMessage, lastCommit[:7], lastTime.Format(time.RFC3339))
```
Comment (pjbgf) [inline controllers/imageupdateautomation_controller.go:323]: `lastCommit[:7]` will panic if lastCommit string is not as long as it assumes it is (e.g. at least 7 characters long). Therefore the if has to confirm that assumption.

**Review by pjbgf — CHANGES_REQUESTED** [review]
> Not too sure about how to test this, happy to get some pointers.

I can't see an easy way to test it, apart from extracting the if-else block of code into its own function and created a unit test just for it.

File: controllers/imageupdateautomation_controller.go (hunk 1 above)
Comment (Nalum) [reply controllers/imageupdateautomation_controller.go:323]: Good catch, I hadn't picked up on this.

**Comment by Nalum** [comment]
> I can't see an easy way to test it, apart from extracting the if-else block of code into its own function and creating a unit test just for it.

I can look at doing this but is seems a little overkill to me.

File: controllers/imageupdateautomation_controller.go [hunk 2]
```diff
@@ -316,15 +316,24 @@ func (r *ImageUpdateAutomationReconciler) Reconcile(ctx context.Context, req ctr
...
+			lpcLen := len(auto.Status.LastPushCommit)
+			message := `%s; last commit "%s" at %s`
+
+			if lpcLen > 7 {
+				lpcLen = 7
+				message = `%s; last commit "%s..." at %s`
+			}
+
+			statusMessage = fmt.Sprintf(message, statusMessage, auto.Status.LastPushCommit[:lpcLen], auto.Status.LastPushTime.Format(time.RFC3339))
+		}
```
Comment (pjbgf) [inline controllers/imageupdateautomation_controller.go:336]: nit: could be simplified for clarity

```suggestion
		if auto.Status.LastPushTime != nil && len(auto.Status.LastPushCommit) >= 7 {
			statusMessage = fmt.Sprintf("%s; last commit %s at %s", statusMessage, auto.Status.LastPushCommit[:7], auto.Status.LastPushTime.Format(time.RFC3339))
		}
```

**Review by pjbgf — COMMENTED** [review]
> I can look at doing this but is seems a little overkill to me.

agreed, besides the ongoing refactoring may help on that front too.

on a side point, I added a nit comment about simplifying the if logic for clarity.

File: controllers/imageupdateautomation_controller.go (hunk 2 above)
Comment (Nalum) [reply controllers/imageupdateautomation_controller.go:336]: I think you've missed a bit of what the new code is doing. The internal `if` on the length of the commit message is only trimming it if it's `>7`, the suggested code will only log the message if the commit message is `>7`. I can add a comment around this given that it's not clear what's happening, if that would help.

Comment (pjbgf) [reply controllers/imageupdateautomation_controller.go:336]: That was intentional, as 7 characters is the short version of the hash used to represent commits. Decreasing the length here will increase the likelihood of collisions. The idea on the suggestion was to also maintain the current behaviour.

As this field represents the SHA1 commits, it should contain either the full 40 characters or 7 characters for the short version. I would assume something went wrong earlier in the process for a non-empty string with length of anything less than 7.

GitHub requires commit hashes to be at least 4 characters long (in some of their GET APIs). But I am not convinced we benefit by decreasing this.

Comment (Nalum) [reply controllers/imageupdateautomation_controller.go:336]: Ah, okay my mistake. I thought this was the commit message not the commit hash. If this is the hash then I'll make that change :+1:

**Review by pjbgf — APPROVED** [review]
Nice one @Nalum, thank you for your contribution. :bow: 

LGTM

**Comment by Nalum** [comment]
Thanks for your help and patience @pjbgf  :+1:

**Comment by stefanprodan** [comment]
@Nalum please squash all commits into a single one.

**Pull request merged**